### PR TITLE
bpo-37324: Remove ABC aliases from collections

### PR DESCRIPTION
**Pull request by hugovk**
Remove deprecated aliases to Abstract Base Classes from the collections module.

This is an updated version of @vstinner's https://github.com/python/cpython/pull/14171, but for Python 3.10.

PR https://github.com/python/cpython/pull/14171 was originally for 3.9, but closed as it was decided to postpone the removal to 3.10 (https://github.com/python/cpython/pull/18545) to give 3rd-party modules like html5lib more time to update.

html5lib updated in https://github.com/html5lib/html5lib-python/pull/403 and released in version 1.1  (June 2020) https://github.com/html5lib/html5lib-python/releases/tag/1.1. 

@vstinner I just put your name in `Doc/whatsnew/3.10.rst` because this is an updated version of your PR, let me know if I should update it.

https://bugs.python.org/issue37324

### CODE REVIEW

**Comment by vstinner** [comment]
Right now, we are currently fighting with many incompatible changes of Python 3.10, so I would prefer to not add another incompatible change in Python 3.10. Maybe we can attempt that in an early stage of Python 3.11. Or maybe the deprecated alias can stay. I don't know.

There are many Fedora Python packages broken by Python 3.10:

* https://bugzilla.redhat.com/show_bug.cgi?id=1890881
* https://bugzilla.redhat.com/buglist.cgi?bug_id=1890881&bug_id_type=anddependson&format=tvp&list_id=11560742

cc @pablogsal (Python 3.10 release manager)

**Comment by rhettinger** [comment]
FWIW, this has been pending for a long time. I don't think more time will help.   Also, if something breaks, the fix is trivially simple.

**Comment by hugovk** [comment]
The original plan was to remove in 3.8:

* https://github.com/python/cpython/pull/5460
* https://bugs.python.org/issue25988

So far postponed twice:

* https://github.com/python/cpython/pull/13409 remove in 3.9
* https://github.com/python/cpython/pull/18545 remove in 3.10

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
See also #20774.

@rhettinger, merge please one of these PRs.

**Comment by rhettinger** [comment]
That isn't up to me.  Try @pablogsal

**Comment by vstinner** [comment]
I'm not against the change, but **I** would prefer to wait for Python 3.11 to push it. As I wrote previously, 3.10 already contains multiple incompatible changes and it's already painful to handle to deal with them. It's inconvenient to get new incompatible changes between alpha releases.

The ABC aliases in collections are deprecated, IMO they can stay for one more release. I propose to wait until 3.11 development cycle starts to push these changes (before Python 3.11.0 alpha1).

---

But I'm also fine if some people disagree and consider that it's still ok to push such changes after 3.10.0 alpha 1.

Hopefully, since the change was included in early Python 3.9 alpha/beta versions, many projects have already been fixed to use collections.abc. Moreover, most projects now dropped Python 2.7 support and so don't need the try/except ImportError dance to support Python 2 and Python 3 in the same code base. For example, pip was blocked by html5lib, but after months of stale PR, html5lib got released with the fix, and pip was updated to get the fix (it vendors a copy of html5lib).

More generally, I rely on @pablogsal wise decision :-) I understood that Pablo will also be Python 3.11 release manager, so he will have to deal with this change anyway :-)

**Comment by pablogsal** [comment]
I have reviewed the PRs and previous discussing and my opinion is that is ok to go ahead with this for Python 3.10 release. This deprecation was scheduled for Python 3.8 and it has been deprecated 8 years ago. We also made an extension in Python 3.9. I think there is not that much value in making another extension. But as a Release Manager, I have no power or authority to decide what to do or what not to do in this situation, so to move forward the best thing is to discuss this in the Steering Council so we can discuss this in the first meeting and unblock the situation.

I will make comment back here with the result.

**Comment by pablogsal** [comment]
We had an SC meeting on 2021-01-11 where this matter was discussed and the SC determined that as this was deprecated since Python 3.3 and was delayed twice (Python3.8 and Python 3.9) and the reason for the extension was to make the Python 2 deprecation a bit easier, we should go ahead and proceed with the removal. We understand and respect the will to not accumulate backwards incompatible changes in the Python 3.10 release but we think that another delay will not be very helpful in this aspect, given as well that the migration from the deprecated code is straightforward enough. 

Thanks to everyone that participated in the discussion and thanks for exposing your views on the matter.

**Review by vstinner — APPROVED** [review]
LGTM. Ok, I respect the SC decision, let's remove it one more time ;-)

**Comment by vstinner** [comment]
Thanks @hugovk, I merged your PR ;-) Welcome to the bright future with no deprecated aliases :-D

**Comment by vstinner** [comment]
This incompatible change was announced at the beginning of the What's New in Python 3.9 document: https://docs.python.org/3.9/whatsnew/3.9.html#you-should-check-for-deprecationwarning-in-your-code

> Aliases to Abstract Base Classes in the collections module, like collections.Mapping alias to collections.abc.Mapping, are kept for one last release for backward compatibility. They will be removed from Python 3.10.

**Pull request merged**
